server/resource/caldav_service.go

```diff
@@ -1,18 +1,545 @@
+		return false, err
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var rowPerm string
+		err := rows.Scan(&perm)
+		if err != nil {
+			log.Error(err, "failed to scan rows")
+			return false, err
+		}
+		if rowPerm == "admin" {
+			return true, nil
+		}
+		if rowPerm == "write" && perm == "admin" {
+			return false, nil
+		} else {
+			return true, nil
+		}
+		if rowPerm == "read" && perm == "read" {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (pa *PGResourceAdapter) GetContent() string {
+	if pa.IsCollection() {
+		return ""
+	}
+	a, err := pa.haveAccess("read")
+	if err != nil {
+		log.Error(err, "failed to get Access ", pa.resourcePath)
+		return ""
+	}
+	if ! a {
+		return ""
+	}
+	rows, err := pa.db.Query("SELECT content FROM calendar WHERE rpath = $1 AND owner_id = $2", pa.resourcePath, pa.UserID)
+	if err != nil {
+		log.Error(err, "failed to fetch content ", pa.resourcePath)
+		return ""
+	}
+	defer rows.Close()
+	var content string
+	for rows.Next() {
+		err = rows.Scan(&content)
+		if err != nil {
+			log.Error(err, "failed to scan ", pa.resourcePath)
+			return ""
+		}
+	}
+	ret, err := base64.StdEncoding.DecodeString(content)
+	if err != nil {
+		log.Error(err, "decode error ", pa.resourcePath)
+		return ""
+	}
+	return string(ret)
+}
+
+func (pa *PGResourceAdapter) GetContentSize() int64 {
+	return int64(len(pa.GetContent()))
+}
+
+func (pa *PGResourceAdapter) IsCollection() bool {
+	return isCollection(pa.resourcePath)
+}
+
+func (pa *PGResourceAdapter) GetModTime() time.Time {
+	a, err := pa.haveAccess("read")
+	if err != nil {
+		log.Error(err, "failed to get Access ", pa.resourcePath)
+		return time.Unix(0, 0)
+	}
+	if ! a {
+		log.Info("failed to get Access ", pa.resourcePath)
+		return time.Unix(0, 0)
+	}
+	rows, err := pa.db.Query("SELECT modified FROM calendar WHERE rpath = $1 AND owner_id = $2", pa.resourcePath, pa.UserID)
+	if err != nil {
+		log.Error(err, "failed to fetch modTime ", pa.resourcePath)
+		return time.Unix(0, 0)
+	}
+	defer rows.Close()
+	var mod time.Time
+	for rows.Next() {
+		err = rows.Scan(&mod)
+		if err != nil {
+			log.Error(err, "failed to scan modTime ", pa.resourcePath)
+			return time.Unix(0, 0)
+		}
+	}
+	return mod
+}
+
+
+func NewCaldavServer(CaldavListenInterface string,  cruds database.DatabaseConnection) *http.Server {
+
+	if err := setupDB(cruds); err != nil {
+		log.Error(err)
+		os.Exit(1)
+	}
+
+	var username, email string
+	var id int64
+	rows, err := cruds.Query("SELECT id, username, email FROM users")
+	if err != nil {
+		log.Error(err)
+		return nil
+	}
+
+	for rows.Next() {
+		err = rows.Scan(&id, &username, &email)
+		if err != nil {
+			log.Error(err)
+			return nil
+		}
+	}
+
+	stg := new(CalDavStorage)
+	stg.cruds    = cruds
+	stg.User   = username
+	stg.UserID = id
+	stg.Email  = email
+	caldav.SetupStorage(stg)
+
 	servermux := http.NewServeMux()
 	servermux.HandleFunc("/caldav", caldav.RequestHandler)
```

Comment:
this need to go to the server.go file where all the path mappings are created. just return handler from this function.